### ISSUE TITLE
Some confusion on mAP of CUKH03 in eval_metrics.py

### ISSUE THREAD

**Issue by kalilili** [KaiyangZhou/deep-person-reid#114]
cmc, AP = 0., 0.
        for repeat_idx in range(num_repeats):
            mask = np.zeros(len(raw_cmc), dtype=np.bool)
            for _, idxs in g_pids_dict.items():
                # randomly sample one image for each gallery person
                rnd_idx = np.random.choice(idxs)
                mask[rnd_idx] = True
            masked_raw_cmc = raw_cmc[mask]
            _cmc = masked_raw_cmc.cumsum()
            _cmc[_cmc > 1] = 1
            cmc += _cmc[:max_rank].astype(np.float32)
            # compute AP
            num_rel = masked_raw_cmc.sum()
            tmp_cmc = masked_raw_cmc.cumsum()
            tmp_cmc = [x / (i+1.) for i, x in enumerate(tmp_cmc)]
            tmp_cmc = np.asarray(tmp_cmc) * masked_raw_cmc
            AP += tmp_cmc.sum() / num_rel

The author calculates the mAP of CUKH03 using single_gallery_shot. In my opinoin, this is superfluous and only necessary for cmc. 
The mAP results calculated in this way are different from others, for instance open-reid. 
https://github.com/Cysu/open-reid/blob/master/reid/evaluation_metrics/ranking.py

**Comment by KaiyangZhou** [comment]
In my opinion, it's hard to judge which implementation is correct. This is because mAP was first introduced in the market1501 dataset to account for multiple ground truths in gallery. Before that, people did not really report mAP on the cuhk03 old setting, only the cmc ranking curve was reported. As far as I know, only few papers reported mAP on the cuhk03 old setting, e.g. [1]. But [1] designed a new setting (767/700 split), which is more popular in recent publications. Most papers that used the old setting only reported the rank results, e.g. [2,3,4]. Therefore, I would suggest to not report the mAP results on the cuhk03 old setting.

It makes sense to me to follow the single-gallery-shot setting for computing the AP for the old cuhk03 setting, I would say.

[1] [Re-ranking Person Re-identification with k-reciprocal Encoding, CVPR'17](https://arxiv.org/abs/1701.08398)
[2] [Multi-Level Factorisation Net for Person Re-Identification, CVPR'18](https://arxiv.org/abs/1803.09132)
[3] [DeepReID: Deep Filter Pairing Neural Network for Person Re-Identification, CVPR'15](https://www.cv-foundation.org/openaccess/content_cvpr_2014/papers/Li_DeepReID_Deep_Filter_2014_CVPR_paper.pdf)
[4] [Learning a Discriminative Null Space for Person Re-identification, CVPR'16](https://arxiv.org/abs/1603.02139)

**Comment by kalilili** [comment]
Thanks for your reply!

**Comment by michuanhaohao** [comment]
@KaiyangZhou Hi，为什么cuhk03新评测协议需要测试num_repeats次，钟准提出新评测协议的时候已经统一了评测方式，和market和duke一样，分为query和gallery，只测一次就可以了，为什么代码里有计算num_repeats次求平均的这个操作

麻烦你了！

**Comment by KaiyangZhou** [comment]
@michuanhaohao that's for cuhk03's old split.

For the new split, you **don't** need `--cuhk03-classic-split`.

**Comment by michuanhaohao** [comment]
@KaiyangZhou OK, thanks very much!

**Comment by BraveApple** [comment]
@KaiyangZhou I have a question about evaluating on CUHK03 when I adopt the new split 767/700 and Euler distance metric. Then, I find that the mAP (55.45%) is higher than the Rank1 accuracy (42.54%), but in my view  the mAP  should be lower than the Rank1 accuracy.

Evaluating cuhk03 ...
The gallery size => [5332]
############ Single-Query Protocal with Gallery Size 5332 #############
=> Computing CMC and MAP
MAP     : 55.45%
Rank-1  : 42.54%
Rank-5  : 71.11%
Rank-10 : 80.01%
############ Single-Query Protocal with Gallery Size 5332 #############

**Comment by KaiyangZhou** [comment]
@BraveApple yes, mAP should be smaller, typically.

Have you made any changes to the code?

**Comment by BraveApple** [comment]
> @BraveApple yes, mAP should be smaller, typically.
> 
> Have you made any changes to the code?
> 
> What command to reproduce your result?

Thanks for your reply. Yes, I have changed codes in the previous result. So I also test the same command with the original code, then I also find the same problem. Here I release my command as follows:
`python train_imgreid_xent_htri.py \
--root ${WPY_DATA}/original_dataset/Person_ReID \
-s cuhk03 \
-t cuhk03 \
--height 256 \
--width 128 \
--optim amsgrad \
--label-smooth \                                                                                                                                                                             
--lr 0.0003 \
--max-epoch 60 \
--stepsize 20 40 \
--train-batch-size 32 \
--test-batch-size 100 \
-a resnet50 \
--save-dir log/resnet50-market-xent \
--gpu-devices 0 \ 
--eval-freq 1 \ 
`
The the log of the fist two epoch is as follows:

`Initializing model: resnet50
Initialized model with pretrained weights from https://download.pytorch.org/models/resnet50-19c8e357.pth
Model size: 23.508 M
=> Start training
Epoch: [1][10/230]      Time 0.190 (0.830)      Data 0.0001 (0.0867)    Xent 7.4725 (6.9198)    Htri 0.4311 (0.1406)    Acc 0.00 (0.62)
Epoch: [1][20/230]      Time 0.189 (0.509)      Data 0.0000 (0.0434)    Xent 7.5521 (7.1287)    Htri 0.0000 (0.0966)    Acc 0.00 (0.62)
Epoch: [1][30/230]      Time 0.189 (0.402)      Data 0.0000 (0.0289)    Xent 6.9127 (7.1953)    Htri 0.9344 (0.1439)    Acc 3.12 (0.52)
Epoch: [1][40/230]      Time 0.189 (0.349)      Data 0.0000 (0.0217)    Xent 7.1984 (7.1646)    Htri 0.0000 (0.1430)    Acc 0.00 (0.39)
Epoch: [1][50/230]      Time 0.189 (0.317)      Data 0.0000 (0.0174)    Xent 6.6812 (7.0995)    Htri 0.0000 (0.1513)    Acc 0.00 (0.31)
Epoch: [1][60/230]      Time 0.188 (0.296)      Data 0.0001 (0.0145)    Xent 6.7666 (7.0453)    Htri 0.0000 (0.1607)    Acc 0.00 (0.26)
Epoch: [1][70/230]      Time 0.188 (0.280)      Data 0.0000 (0.0124)    Xent 6.6527 (7.0036)    Htri 0.0000 (0.1494)    Acc 0.00 (0.22)
Epoch: [1][80/230]      Time 0.191 (0.269)      Data 0.0000 (0.0109)    Xent 6.7733 (6.9691)    Htri 0.0000 (0.1332)    Acc 0.00 (0.20)
Epoch: [1][90/230]      Time 0.191 (0.260)      Data 0.0000 (0.0097)    Xent 6.8919 (6.9442)    Htri 0.0000 (0.1259)    Acc 0.00 (0.21)
Epoch: [1][100/230]     Time 0.191 (0.253)      Data 0.0000 (0.0087)    Xent 6.6793 (6.9224)    Htri 0.0211 (0.1144)    Acc 0.00 (0.19)
Epoch: [1][110/230]     Time 0.190 (0.248)      Data 0.0001 (0.0079)    Xent 6.7115 (6.9029)    Htri 0.3849 (0.1126)    Acc 0.00 (0.20)
Epoch: [1][120/230]     Time 0.190 (0.243)      Data 0.0000 (0.0073)    Xent 6.6382 (6.8857)    Htri 0.0000 (0.1078)    Acc 3.12 (0.21)
Epoch: [1][130/230]     Time 0.194 (0.239)      Data 0.0000 (0.0067)    Xent 6.7289 (6.8722)    Htri 0.0000 (0.1004)    Acc 0.00 (0.22)
Epoch: [1][140/230]     Time 0.208 (0.236)      Data 0.0000 (0.0062)    Xent 6.6636 (6.8596)    Htri 0.3357 (0.0980)    Acc 0.00 (0.20)
Epoch: [1][150/230]     Time 0.212 (0.234)      Data 0.0000 (0.0058)    Xent 6.7554 (6.8496)    Htri 0.0099 (0.0917)    Acc 0.00 (0.19)
Epoch: [1][160/230]     Time 0.212 (0.233)      Data 0.0000 (0.0055)    Xent 6.6771 (6.8394)    Htri 0.0000 (0.0873)    Acc 0.00 (0.18)
Epoch: [1][170/230]     Time 0.212 (0.232)      Data 0.0000 (0.0051)    Xent 6.7039 (6.8300)    Htri 0.0000 (0.0830)    Acc 0.00 (0.17)
Epoch: [1][180/230]     Time 0.212 (0.231)      Data 0.0000 (0.0049)    Xent 6.6755 (6.8214)    Htri 0.0078 (0.0821)    Acc 0.00 (0.16)
Epoch: [1][190/230]     Time 0.214 (0.230)      Data 0.0000 (0.0046)    Xent 6.7001 (6.8137)    Htri 0.0000 (0.0796)    Acc 0.00 (0.15)
Epoch: [1][200/230]     Time 0.212 (0.229)      Data 0.0000 (0.0044)    Xent 6.6619 (6.8067)    Htri 0.0000 (0.0766)    Acc 0.00 (0.14)
Epoch: [1][210/230]     Time 0.212 (0.228)      Data 0.0000 (0.0042)    Xent 6.6711 (6.8005)    Htri 0.2091 (0.0746)    Acc 0.00 (0.13)
Epoch: [1][220/230]     Time 0.212 (0.227)      Data 0.0001 (0.0040)    Xent 6.6668 (6.7945)    Htri 0.0000 (0.0717)    Acc 0.00 (0.13)
Epoch: [1][230/230]     Time 0.212 (0.227)      Data 0.0000 (0.0038)    Xent 6.6606 (6.7893)    Htri 0.0417 (0.0697)    Acc 0.00 (0.14)
=> Test
Evaluating cuhk03 ...
Extracted features for query set, obtained 1400-by-2048 matrix
Extracted features for gallery set, obtained 5332-by-2048 matrix
=> BatchTime(s)/BatchSize(img): 0.062/100
Computing CMC and mAP
Results ----------
mAP: 0.8%
CMC curve
Rank-1  : 0.5%
Rank-5  : 1.6%
Rank-10 : 2.2%
Rank-20 : 3.6%
------------------
Checkpoint saved to "log/resnet50-market-xent/model.pth.tar-1"
Epoch: [2][10/230]      Time 0.191 (0.302)      Data 0.0001 (0.1068)    Xent 6.6390 (6.6306)    Htri 0.0290 (0.0116)    Acc 0.00 (0.31)
Epoch: [2][20/230]      Time 0.191 (0.247)      Data 0.0000 (0.0534)    Xent 6.6193 (6.6263)    Htri 0.0000 (0.0137)    Acc 0.00 (0.31)
Epoch: [2][30/230]      Time 0.191 (0.228)      Data 0.0000 (0.0356)    Xent 6.6276 (6.6271)    Htri 0.0237 (0.0206)    Acc 0.00 (0.31)
Epoch: [2][40/230]      Time 0.191 (0.219)      Data 0.0001 (0.0267)    Xent 6.6028 (6.6225)    Htri 0.0000 (0.0173)    Acc 0.00 (0.31)
Epoch: [2][50/230]      Time 0.192 (0.214)      Data 0.0001 (0.0214)    Xent 6.6417 (6.6239)    Htri 0.0000 (0.0205)    Acc 0.00 (0.25)
Epoch: [2][60/230]      Time 0.198 (0.211)      Data 0.0001 (0.0178)    Xent 6.6367 (6.6228)    Htri 0.0295 (0.0199)    Acc 0.00 (0.26)
Epoch: [2][70/230]      Time 0.211 (0.210)      Data 0.0001 (0.0153)    Xent 6.6510 (6.6236)    Htri 0.0337 (0.0184)    Acc 0.00 (0.22)
Epoch: [2][80/230]      Time 0.214 (0.210)      Data 0.0001 (0.0134)    Xent 6.6219 (6.6208)    Htri 0.0000 (0.0180)    Acc 0.00 (0.20)
Epoch: [2][90/230]      Time 0.214 (0.211)      Data 0.0000 (0.0119)    Xent 6.5042 (6.6196)    Htri 0.0000 (0.0180)    Acc 3.12 (0.21)
Epoch: [2][100/230]     Time 0.214 (0.211)      Data 0.0000 (0.0107)    Xent 6.5574 (6.6171)    Htri 0.0000 (0.0164)    Acc 0.00 (0.19)
Epoch: [2][110/230]     Time 0.214 (0.211)      Data 0.0001 (0.0098)    Xent 6.6050 (6.6149)    Htri 0.0000 (0.0165)    Acc 0.00 (0.17)
Epoch: [2][120/230]     Time 0.213 (0.212)      Data 0.0000 (0.0089)    Xent 6.5389 (6.6108)    Htri 0.0874 (0.0204)    Acc 0.00 (0.18)
Epoch: [2][130/230]     Time 0.213 (0.212)      Data 0.0001 (0.0083)    Xent 6.6088 (6.6097)    Htri 0.0224 (0.0221)    Acc 0.00 (0.17)
Epoch: [2][140/230]     Time 0.213 (0.212)      Data 0.0000 (0.0077)    Xent 6.5482 (6.6102)    Htri 0.0197 (0.0221)    Acc 0.00 (0.16)
Epoch: [2][150/230]     Time 0.212 (0.212)      Data 0.0001 (0.0072)    Xent 6.6430 (6.6106)    Htri 0.0000 (0.0281)    Acc 0.00 (0.15)
Epoch: [2][160/230]     Time 0.213 (0.212)      Data 0.0000 (0.0067)    Xent 6.6326 (6.6118)    Htri 0.0000 (0.0277)    Acc 0.00 (0.16)
Epoch: [2][170/230]     Time 0.214 (0.212)      Data 0.0001 (0.0063)    Xent 6.5846 (6.6103)    Htri 0.0000 (0.0261)    Acc 0.00 (0.15)
Epoch: [2][180/230]     Time 0.214 (0.212)      Data 0.0000 (0.0060)    Xent 6.6127 (6.6109)    Htri 0.0000 (0.0264)    Acc 0.00 (0.14)
Epoch: [2][190/230]     Time 0.213 (0.212)      Data 0.0000 (0.0057)    Xent 6.6002 (6.6087)    Htri 0.0000 (0.0255)    Acc 0.00 (0.18)
Epoch: [2][200/230]     Time 0.214 (0.212)      Data 0.0001 (0.0054)    Xent 6.6293 (6.6064)    Htri 0.0000 (0.0268)    Acc 0.00 (0.22)
Epoch: [2][210/230]     Time 0.214 (0.212)      Data 0.0001 (0.0051)    Xent 6.6488 (6.6056)    Htri 0.0000 (0.0255)    Acc 0.00 (0.22)
Epoch: [2][220/230]     Time 0.214 (0.212)      Data 0.0001 (0.0049)    Xent 6.4101 (6.6027)    Htri 0.0000 (0.0251)    Acc 3.12 (0.23)
Epoch: [2][230/230]     Time 0.212 (0.212)      Data 0.0000 (0.0047)    Xent 6.5913 (6.5976)    Htri 0.0000 (0.0278)    Acc 0.00 (0.23)
=> Test
Evaluating cuhk03 ...
Extracted features for query set, obtained 1400-by-2048 matrix
Extracted features for gallery set, obtained 5332-by-2048 matrix
=> BatchTime(s)/BatchSize(img): 0.005/100
Computing CMC and mAP
Results ----------
mAP: 1.6%
CMC curve
Rank-1  : 1.1%
Rank-5  : 3.3%
Rank-10 : 5.5%
Rank-20 : 8.1%
------------------
Checkpoint saved to "log/resnet50-market-xent/model.pth.tar-2"
`

**Comment by KaiyangZhou** [comment]
perhaps wrap your log with ``` for better readability?

**Comment by KaiyangZhou** [comment]
@BraveApple  I tested with the command you gave. I got similar results where mAP is slightly higher than rank-1 at early epochs. With the model becoming better, rank-1 increases over time and the ratio compared with mAP is what we expect.

I think this is normal because at early stage the rank-1 is very low, so when computing AP, `P(k) * rel(k)` will be zeros for small `k`. `\sum P(k)*rel(k)/N `is thus possible to be larger than rank-1 in this case, but will not differ substantially I guess. (the AP metric is referred to https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Mean_average_precision)

I don't see there is any bug in the eval code.

**Comment by Adorablepet** [comment]
@KaiyangZhou Could I understand that writing commands in the label mode?
                         cuhk03's new split：--cuhk03-labeled
                         cuhk03's old split: --cuhk03-classic-split --use-metric-cuhk03 --cuhk03-labeled
But I don’t understand a bit, that is why the number of camerars is two in dataset statistics form.The number of cameras in this data set should be five.Thanks.Split_id should set a specific value in cuhk03's old split, or 0~19(cuhk03(classic split)).This does not understand，thanks.

**Comment by KaiyangZhou** [comment]
> @KaiyangZhou Could I understand that writing commands in the label mode?
> cuhk03's new split：--cuhk03-labeled
> cuhk03's old split: --cuhk03-classic-split --use-metric-cuhk03 --cuhk03-labeled

Yes. This is explained in [DATASETS.md](https://github.com/KaiyangZhou/deep-person-reid/blob/master/DATASETS.md).

> But I don’t understand a bit, that is why the number of camerars is two in dataset statistics form.The number of cameras in this data set should be five.Thanks

If I recall correctly, the image name given by cuhk03.mat (the raw matlab file) does not clearly indicate the camera id.

> Split_id should set a specific value in cuhk03's old split, or 0~19(cuhk03(classic split)).This does not understand，thanks.

Yes, you need to do train+test for **20 times** by varying `--split-id=0` to `--split-id=19`.

**Comment by Adorablepet** [comment]
@KaiyangZhou This is the configuration parameter when I write the script about split-id.
                         cuhk03's new split:--split-id=0 
                         cuhk03's old split: --split-id=19
I don't know how to show it（0~19） in the script.Thanks.

**Comment by KaiyangZhou** [comment]
@Adorablepet 
for the new split, you don't need to specify `--split-id` as its default is 0.

for the old split, you need to run 20 scripts, the first one with `--split-id 0`, the second one with `--split-id 1`, ..., and the last one with `--split-id 19`.
